[executorch][core] NamedDataMap interface

lucylq opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #7900
* __->__ #7763

Add NamedDataMap interface to runtime.

Differential Revision: [D66834552](https://our.internmc.facebook.com/intern/diff/D66834552/)

lucylq committed:
[executorch][core] NamedDataMap interface

Add NamedDataMap interface to runtime.

Differential Revision: [D66834552](https://our.internmc.facebook.com/intern/diff/D66834552/)

[ghstack-poisoned]

JacobSzwejbka commented on runtime/core/named_data_map.h:
```diff
@@ -0,0 +1,47 @@
+   */
+  ET_NODISCARD virtual Result<Span<const uint8_t>> get_data(
+      const char* fqn) const = 0;
+};
```
Can you add the functions to get the list of fqns?

dbort commented on runtime/core/named_data_map.h:
```diff
@@ -0,0 +1,47 @@
+   * @param fqn Fully qualified name of the tensor.
+   * @return Result containing a span of uint8_t representing the tensor data.
+   */
+  ET_NODISCARD virtual Result<Span<const uint8_t>> get_data(
```
This needs to return a FreeableBuffer, otherwise there's no way to manage the data. Span doesn't own its data, so anything returned here would be forced to leak memory.

dbort commented on runtime/core/named_data_map.h:
```diff
@@ -0,0 +1,47 @@
+namespace runtime {
+
+/**
+ * Interface to access and retrieve data via name from a loaded data file.
```
NDM has no opinion about where the data comes from, so "loaded data file" is too specific. Pointing to the example is ok, though.

dbort commented on runtime/core/named_data_map.h:
```diff
@@ -0,0 +1,47 @@
+   * Get tensor metadata by fully qualified name (FQN).
+   *
+   * @param fqn Fully qualified name of the tensor.
+   * @return Result containing a pointer to the metadata.
```
Doesn't appear to be a pointer

dbort commented on runtime/core/named_data_map.h:
```diff
@@ -0,0 +1,47 @@
+ * Interface to access and retrieve data via name from a loaded data file.
+ * See executorch/extension/flat_tensor/ for an example.
+ */
+class NamedDataMap {
```
Please annotate this as ET_EXPERIMENTAL so that it doesn't become part of our supported API just yet.

lucylq committed:
Update on "[executorch][core] NamedDataMap interface"


Add NamedDataMap interface to runtime.

Differential Revision: [D66834552](https://our.internmc.facebook.com/intern/diff/D66834552/)

[ghstack-poisoned]

JacobSzwejbka commented on runtime/core/named_data_map.h:
```diff
@@ -0,0 +1,62 @@
+   * @param fqn Fully qualified name of the tensor.
+   * @return Result containing a FreeableBuffer with the tensor data.
+   */
+  ET_NODISCARD virtual Result<FreeableBuffer> get_data(
```
Ok last request from my side is for a get_data_into method. This lets us avoid a useless load into the freeableBuffer and lets us just load directly into the buffer we are targeting for mutation. Primary usage is training, but it would also be used if someone put serialized mutable state like say an initial cache into the .ptd

JacobSzwejbka replied:
https://github.com/pytorch/executorch/blob/main/runtime/executor/program.h#L251
and
https://github.com/pytorch/executorch/blob/main/runtime/core/data_loader.h#L107

for code pointers of where we do this in .pte loading. I think this would be roughly parallel to the program.h code.

lucylq committed:
Update on "[executorch][core] NamedDataMap interface"


Add NamedDataMap interface to runtime.

Differential Revision: [D66834552](https://our.internmc.facebook.com/intern/diff/D66834552/)

[ghstack-poisoned]

lucylq committed:
Update on "[executorch][core] NamedDataMap interface"


Add NamedDataMap interface to runtime.

Differential Revision: [D66834552](https://our.internmc.facebook.com/intern/diff/D66834552/)

[ghstack-poisoned]

dbort commented on runtime/core/named_data_map.h:
```diff
@@ -0,0 +1,74 @@
+      const char* fqn) const = 0;
+
+  /**
+   * Loads data corresponding to the fqn into the provided buffer.
```
Would be good to tell users how to figure out how big a buffer to provide (get metadata and call .nbytes()?)

dbort commented on runtime/core/named_data_map.h:
```diff
@@ -0,0 +1,74 @@
+   * @param size The number of bytes to load.
+   * @param buffer The buffer to load the data into. Must point to at least
+   * `size` bytes of memory.
+   * @return An error code on if the load was successful.
```
Not sure what this means.

Also, what happens if the size is too small or too large?

lucylq replied:
The NamedDataMap implementation should throw an error if the size is too small. Thanks for the suggestion to return Result<size_t>, this can provide information on if the buffer was too large.

dbort commented on runtime/core/named_data_map.h:
```diff
@@ -0,0 +1,74 @@
+   * `size` bytes of memory.
+   * @return An error code on if the load was successful.
+   */
+  ET_NODISCARD virtual Error
```
Consider returning `Result<size_t>`, returning the number of bytes written on success. Then, users can provide a too-big buffer but still know how many bytes are actually valid.

lucylq replied:
thanks!

dbort commented on runtime/core/named_data_map.h:
```diff
@@ -0,0 +1,74 @@
+   *
+   * @return Result containing the number of keys.
+   */
+  ET_NODISCARD virtual Result<int> get_num_keys() const = 0;
```
Please use size_t for this
```suggestion
  ET_NODISCARD virtual Result<size_t> get_num_keys() const = 0;
```

dbort commented on runtime/core/named_data_map.h:
```diff
@@ -0,0 +1,74 @@
+   * @param index The index of the key to retrieve.
+   * @return Result containing the key at the given index.
+   */
+  ET_NODISCARD virtual Result<const char*> get_key(int index) const = 0;
```
```suggestion
  ET_NODISCARD virtual Result<const char*> get_key(size_t index) const = 0;
```

dbort commented on runtime/core/named_data_map.h:
```diff
@@ -0,0 +1,74 @@
+   * Get the key at the given index.
+   *
+   * @param index The index of the key to retrieve.
+   * @return Result containing the key at the given index.
```
Please mention that the returned pointer is only valid for the lifetime of the NDM.

dbort reviewed (changes requested):
Looking good!

lucylq committed:
Update on "[executorch][core] NamedDataMap interface"


Add NamedDataMap interface to runtime.

Differential Revision: [D66834552](https://our.internmc.facebook.com/intern/diff/D66834552/)

[ghstack-poisoned]

dbort commented on runtime/core/named_data_map.h:
```diff
@@ -0,0 +1,76 @@
+   * size of the tensor for a given fqn.
+   * @param buffer The buffer to load the data into. Must point to at least
+   * `size` bytes of memory.
+   * @return Result containing the number of bytes written on success.
```
Please clarify if this will fail or truncate if `size` is too small

dbort commented on runtime/core/named_data_map.h:
```diff
@@ -0,0 +1,76 @@
+   * @return Result containing the number of bytes written on success.
+   */
+  ET_NODISCARD virtual Result<size_t>
+  load_data_into(const char* fqn, size_t size, void* buffer) const = 0;
```
nit: traditionally (like in memcpy) the buffer comes before the size

```suggestion
  load_data_into(const char* fqn, void* buffer, size_t size) const = 0;
```

dbort commented on runtime/core/named_data_map.h:
```diff
@@ -0,0 +1,76 @@
+   *
+   * @return Result containing the number of keys.
+   */
+  ET_NODISCARD virtual Result<size_t> get_num_keys() const = 0;
```
These two methods use "key" to refer to the strings, but the other methods use "fqn". It'd be better to use a single term for the entire interface.

Since NDM is fairly generic, I like "key" better, since we might use an NDM to look up data that is keyed by strings that aren't technically pytorch FQNs. But the most important thing is that the API is consistent.

dbort reviewed (approved):
Some final minor things but this looks good!

lucylq committed:
Update on "[executorch][core] NamedDataMap interface"


Add NamedDataMap interface to runtime.

Differential Revision: [D66834552](https://our.internmc.facebook.com/intern/diff/D66834552/)

[ghstack-poisoned]
